File: app/src/main/java/tw/com/louis383/coffeefinder/mainpage/MainActivity.java
package tw.com.louis383.coffeefinder.mainpage;

import android.Manifest;
import android.animation.Animator;
import android.content.Context;
import android.content.Intent;
import android.content.IntentSender;
import android.content.pm.PackageManager;
import android.location.Location;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.provider.Settings;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.BottomSheetBehavior;
import android.support.design.widget.CoordinatorLayout;
import android.support.design.widget.FloatingActionButton;
import android.support.design.widget.Snackbar;
import android.support.v4.app.ActivityCompat;
import android.support.v4.app.Fragment;
import android.support.v4.content.ContextCompat;
import android.support.v4.widget.NestedScrollView;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.AppCompatRatingBar;
import android.util.Log;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.animation.DecelerateInterpolator;
import android.view.animation.OvershootInterpolator;
import android.widget.Button;
import android.widget.ProgressBar;
import android.widget.TextView;
import com.google.android.gms.common.api.ResolvableApiException;
import com.google.android.gms.location.FusedLocationProviderClient;
import com.google.android.gms.location.LocationServices;
import com.google.android.gms.maps.model.LatLng;
import java.util.List;
import javax.inject.Inject;
import tw.com.louis383.coffeefinder.CoffeeTripApplication;
import tw.com.louis383.coffeefinder.R;
import tw.com.louis383.coffeefinder.about.AboutActivity;
import tw.com.louis383.coffeefinder.list.ListFragment;
import tw.com.louis383.coffeefinder.maps.MapsClickHandler;
import tw.com.louis383.coffeefinder.maps.MapsFragment;
import tw.com.louis383.coffeefinder.model.CoffeeShopListManager;
import tw.com.louis383.coffeefinder.model.domain.CoffeeShop;
import tw.com.louis383.coffeefinder.utils.Utils;
import tw.com.louis383.coffeefinder.viewmodel.CoffeeShopViewModel;

/**
 * Created by louis383 on 2017/2/17.
 */

public class MainActivity extends AppCompatActivity implements MainPresenter.MainView,  MapsClickHandler, ListFragment.Callback {
    private static final int LOCATION_PERMISSION_REQUEST = 0;
    private static final int LOCATION_MANUAL_ENABLE = 1;
    private static final int LOCATION_SETTING_RESOLUTION = 2;
    private static final int INTERNET_REQUEST = 3;

    private MainPresenter presenter;
    private Snackbar snackbar;
    private BottomSheetBehavior bottomSheetBehavior;

    // Main Content
    private CoordinatorLayout rootView;
    private NestedScrollView bottomSheet;
    private View shadow;

    // Bottom Sheet
    private FloatingActionButton navigationFab;
    private TextView bottomSheetTitle;
    private TextView bottomSheetDistance;
    private TextView bottomSheetWifiScore;
    private TextView bottomSheetSeatScore;
    private TextView bottomSheetLimitedTime;
    private TextView bottomSheetSocket;
    private TextView bottomSheetStandingDesk;
    private TextView bottomSheetOpenTime;
    private TextView bottomSheetWebsite;
    private TextView bottomSheetMrt;
    private AppCompatRatingBar bottomSheetExpensebar;
    private ProgressBar bottomSheetWifiQuality;
    private ProgressBar bottomSheetSeatQuality;
    private Button bottomSheetNavigate;
    private Button bottomSheetShare;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        setTheme(R.style.AppTheme);
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        ((CoffeeTripApplication) getApplication()).getAppComponent().inject(this);

        // Main Content
        rootView = (CoordinatorLayout) findViewById(R.id.main_rootview);
        bottomSheet = (NestedScrollView) findViewById(R.id.main_bottom_sheet);
        shadow = findViewById(R.id.main_shadow);

        // Bottom Sheet
        navigationFab = (FloatingActionButton) findViewById(R.id.main_fab);
        bottomSheetTitle = (TextView) findViewById(R.id.detail_view_title);
        bottomSheetDistance = (TextView) findViewById(R.id.detail_view_distance);
        bottomSheetWifiScore = (TextView) findViewById(R.id.detail_view_wifi_score);
        bottomSheetSeatScore = (TextView) findViewById(R.id.detail_view_seat_score);
        bottomSheetLimitedTime = (TextView) findViewById(R.id.detail_view_limited_time);
        bottomSheetSocket = (TextView) findViewById(R.id.detail_view_socket);
        bottomSheetStandingDesk = (TextView) findViewById(R.id.detail_view_standing_desk);
        bottomSheetOpenTime = (TextView) findViewById(R.id.detail_view_opentime);
        bottomSheetWebsite = (TextView) findViewById(R.id.detail_view_website);
        bottomSheetMrt = (TextView) findViewById(R.id.detail_view_mrt);
        bottomSheetExpensebar = (AppCompatRatingBar) findViewById(R.id.detail_view_expense);
        bottomSheetWifiQuality = (ProgressBar) findViewById(R.id.detail_view_wifi_quality);
        bottomSheetSeatQuality = (ProgressBar) findViewById(R.id.detail_view_seat_quality);
        bottomSheetNavigate = (Button) findViewById(R.id.detail_view_button_navigate);
        bottomSheetShare = (Button) findViewById(R.id.detail_view_button_share);

        init();
        bottomSheetBehavior = BottomSheetBehavior.from(bottomSheet);
        bottomSheetBehavior.setState(BottomSheetBehavior.STATE_HIDDEN);

        presenter.attachView(this);
        presenter.addLifecycleOwner(this);
        presenter.setBottomSheetBehavior(bottomSheetBehavior);

        bottomSheetNavigate.setOnClickListener(v -> presenter.prepareNavigation());
        bottomSheetShare.setOnClickListener(v -> presenter.share(MainActivity.this));
    }

    private void init() {
        MapsFragment mapsFragment = MapsFragment.newInstance();
        mapsFragment.setRetainInstance(true);
        mapsFragment.setMapClickHandler(this);

        getSupportFragmentManager().beginTransaction()
                .replace(R.id.main_container, mapsFragment)
                .commit();
    }

    @Inject
    public void initPresenter(CoffeeShopListManager coffeeShopListManager) {
        FusedLocationProviderClient fusedLocationProviderClient = LocationServices
                .getFusedLocationProviderClient(this);
        presenter = new MainPresenter(coffeeShopListManager, fusedLocationProviderClient);
    }

    @Override
    public Context getActivityContext() {
        return this;
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        switch (requestCode) {
            case LOCATION_PERMISSION_REQUEST:
                if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                    presenter.requestUserLocation(true);
                } else {
                    if (ActivityCompat.shouldShowRequestPermissionRationale(this, Manifest.permission.ACCESS_FINE_LOCATION)) {
                        new AlertDialog.Builder(this)
                                .setMessage(Utils.getResourceString(this, R.string.request_location))
                                .setPositiveButton(Utils.getResourceString(this, R.string.dialog_auth),
                                        (dialog, which) -> requestLocationPermission())
                                .setNegativeButton(Utils.getResourceString(this, R.string.dialog_cancel),
                                        (dialog, which) -> showPermissionNeedSnackBar())
                                .create().show();
                    } else {
                        showPermissionNeedSnackBar();
                        String appName = Utils.getResourceString(this, R.string.app_name);
                        String permissionName = Utils.getResourceString(this, R.string.auth_location);

                        new AlertDialog.Builder(this)
                                .setTitle(Utils.getResourceString(this, R.string.dialog_auth))
                                .setMessage(getResources().getString(R.string.auth_yourself, appName, permissionName))
                                .setPositiveButton(Utils.getResourceString(this, R.string.auto_go),
                                        (dialog, which) -> {
                                            openApplicationSetting();
                                        })
                                .setNegativeButton(Utils.getResourceString(this, R.string.dialog_cancel), (dialog, which) -> {})
                                .create().show();
                    }
                }

                break;
            default:
                super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        switch (requestCode) {
            case INTERNET_REQUEST:
                if (isInternetAvailable()) {
                    if (checkLocationPermission()) {
                        forceRequestCoffeeShop();
                    } else {
                        showPermissionNeedSnackBar();
                    }
                } else {
                    requestInternetConnection();
                }
                break;
            case LOCATION_MANUAL_ENABLE:
                if (checkLocationPermission()) {
                    forceRequestCoffeeShop();
                    if (snackbar != null && snackbar.isShown()) {
                        snackbar.dismiss();
                    }
                } else {
                    showPermissionNeedSnackBar();
                }
                break;
            case LOCATION_SETTING_RESOLUTION:
                if (resultCode == RESULT_OK) {
                    presenter.requestUserLocation(true);
                } else {
                    snackbar = Snackbar.make(rootView, R.string.high_accuracy_recommand, Snackbar.LENGTH_LONG);
                    snackbar.show();
                }
                break;
            default:
                super.onActivityResult(requestCode, resultCode, data);
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater menuInflater = getMenuInflater();
        menuInflater.inflate(R.menu.main_activity, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_about:
                // Go to about page!
                Intent intent = new Intent(this, AboutActivity.class);
                startActivity(intent);
                break;
        }

        return super.onOptionsItemSelected(item);
    }

    private void forceRequestCoffeeShop() {
        presenter.requestUserLocation(true);
    }

    @Override
    public boolean checkLocationPermission() {
        return ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED;
    }

    @Override
    public void requestLocationPermission() {
        String[] permissions = { Manifest.permission.ACCESS_FINE_LOCATION };
        ActivityCompat.requestPermissions(this, permissions, LOCATION_PERMISSION_REQUEST);
    }

    @Override
    public void locationSettingNeedsResolution(ResolvableApiException resolvable) {
        try {
            resolvable.startResolutionForResult(this, LOCATION_SETTING_RESOLUTION);
        } catch (IntentSender.SendIntentException e) {
            Log.e("MainActivity", Log.getStackTraceString(e));
        }
    }

    @Override
    public void showServiceUnavailableMessage() {
        String message = getResources().getString(R.string.service_unavailable);
        makeSnackBar(message, true);
    }

    @Override
    public void makeSnackBar(String message, boolean infinity) {
        int duration = infinity ? Snackbar.LENGTH_INDEFINITE : Snackbar.LENGTH_LONG;
        snackbar = Snackbar.make(rootView, message, duration);
        snackbar.show();
    }

    @Override
    public void setStatusBarDarkIndicator() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            getWindow().getDecorView().setSystemUiVisibility(View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR);
        }
    }

    @Override
    public void requestInternetConnection() {
        new AlertDialog.Builder(this)
            .setTitle(getResourceString(R.string.internet_request_title))
            .setMessage(getResourceString(R.string.internet_request_message))
            .setPositiveButton(getResourceString(R.string.auto_go), (dialog, which) -> {
                Intent intent = new Intent(Settings.ACTION_WIFI_SETTINGS);
                startActivityForResult(intent, INTERNET_REQUEST);
            }).create().show();
    }

    public Location getCurrentLocation() {
        return presenter.getCurrentLocation();
    }

    @Override
    public void moveCameraToCurrentPosition(LatLng latLng) {
        MapsFragment mapsFragment = getMapFragment();
        if (mapsFragment != null) {
            mapsFragment.moveCamera(latLng, MapsFragment.ZOOM_RATE);
        }
    }

    @Override
    public void onCoffeeShopFetched(List<CoffeeShop> coffeeShops) {
        MapsFragment mapsFragment = getMapFragment();
        if (mapsFragment != null) {
            mapsFragment.prepareCoffeeShops(coffeeShops);
        }
    }

    @Override
    public boolean isApplicationInstalled(String packageName) {
        PackageManager packageManager = getPackageManager();
        try {
            packageManager.getPackageInfo(packageName, PackageManager.GET_ACTIVITIES);
            return true;
        } catch (PackageManager.NameNotFoundException e) {
            return false;
        }
    }

    @Override
    public void navigateToLocation(Intent intent) {
        startActivity(intent);
    }

    @Override
    public void showNeedsGoogleMapMessage() {
        String message = getResources().getString(R.string.googlemap_not_install);
        makeSnackBar(message, false);
    }

    @Override
    public void shareCoffeeShop(Intent shareIntent) {
        String title = getResourceString(R.string.share_title);
        startActivity(Intent.createChooser(shareIntent, title));
    }

    private void showPermissionNeedSnackBar() {
        if (snackbar == null) {
            snackbar = Snackbar.make(rootView, R.string.permission_needed, Snackbar.LENGTH_INDEFINITE);
            snackbar.setAction(R.string.dialog_auth, v -> openApplicationSetting());
        }
        snackbar.show();
    }

    private void openApplicationSetting() {
        Intent intent = new Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS);
        Uri uri = Uri.fromParts("package", getPackageName(), null);
        intent.setData(uri);
        startActivityForResult(intent, LOCATION_MANUAL_ENABLE);
    }

    private String getResourceString(int stringId) {
        return getResources().getString(stringId);
    }

    @Override
    public void showBottomSheetDetailView(CoffeeShopViewModel viewModel) {
        if (bottomSheet != null && bottomSheetBehavior != null) {
            bottomSheetBehavior.setState(BottomSheetBehavior.STATE_COLLAPSED);

            bottomSheetTitle.setText(viewModel.getShopName());
            bottomSheetDistance.setText(viewModel.getDistances());
            bottomSheetExpensebar.setRating(viewModel.getCheapPoints());

            bottomSheetWifiQuality.setProgress((int) viewModel.getWifiPoints() * 20);
            bottomSheetWifiScore.setText(String.valueOf(viewModel.getWifiPoints()));
            bottomSheetSeatQuality.setProgress((int) viewModel.getSeatPoints() * 20);
            bottomSheetSeatScore.setText(String.valueOf(viewModel.getSeatPoints()));

            bottomSheetWebsite.setText(viewModel.getWebsiteURL(this));
            bottomSheetOpenTime.setText(viewModel.getOpenTimes(this));
            bottomSheetMrt.setText(viewModel.getMrtInfo(this));

            bottomSheetLimitedTime.setText(viewModel.getLimitTimeString(this));
            bottomSheetSocket.setText(viewModel.getSocketString(this));
            bottomSheetStandingDesk.setText(viewModel.getStandingDeskString(this));
        }
    }

    @Override
    public void showFab(boolean show) {
        if (show) {
            navigationFab.animate().scaleX(1).scaleY(1).setDuration(300)
                    .setListener(new Animator.AnimatorListener() {
                        @Override
                        public void onAnimationStart(Animator animation) {
                            if (!navigationFab.isShown()) {
                                navigationFab.setVisibility(View.VISIBLE);
                            }
                        }
                        @Override
                        public void onAnimationEnd(Animator animation) {}
                        @Override
                        public void onAnimationCancel(Animator animation) {}
                        @Override
                        public void onAnimationRepeat(Animator animation) {}
                    })
                    .setInterpolator(new OvershootInterpolator())
                    .start();
        } else {
            navigationFab.animate().scaleX(0).scaleY(0).setDuration(200).setInterpolator(new DecelerateInterpolator()).start();
        }
    }

    @Override
    public void setFloatingActionButtonEnable(boolean enable) {
        if (!enable) {
            navigationFab.setOnClickListener(null);
            navigationFab.setVisibility(View.GONE);
        } else {
            navigationFab.setOnClickListener(v -> presenter.prepareNavigation());
        }

        navigationFab.setEnabled(enable);
        navigationFab.setClickable(enable);
    }

    @Override
    public void setShadowAlpha(float offset) {
        if (offset > 0.0f) {
            float alpha = offset * 0.4f;
            shadow.setAlpha(alpha);
        }

        shadow.setVisibility(offset > 0.0f ? View.VISIBLE : View.GONE);
    }

    @Override
    public boolean isInternetAvailable() {
        ConnectivityManager manager = (ConnectivityManager) getSystemService(Context.CONNECTIVITY_SERVICE);
        NetworkInfo networkInfo = manager.getActiveNetworkInfo();
        return networkInfo != null && networkInfo.isConnected() && networkInfo.isAvailable();
    }

    @Override
    public void onBackPressed() {
        if (bottomSheetBehavior != null) {
            if (bottomSheetBehavior.getState() == BottomSheetBehavior.STATE_EXPANDED) {
                bottomSheet.fullScroll(View.FOCUS_UP);
                bottomSheetBehavior.setState(BottomSheetBehavior.STATE_COLLAPSED);
                return;
            } else if (bottomSheetBehavior.getState() == BottomSheetBehavior.STATE_COLLAPSED) {
                showFab(false);
                bottomSheetBehavior.setState(BottomSheetBehavior.STATE_HIDDEN);
                return;
            }
        }

        super.onBackPressed();
    }

    //region MapsClickHandler
    @Override
    public void onMapClicked() {
        if (bottomSheetBehavior != null) {
            showFab(false);
            bottomSheetBehavior.setState(BottomSheetBehavior.STATE_HIDDEN);
        }
    }

    @Override
    public void onMarkerClicked(CoffeeShop coffeeShop) {
        presenter.setLastTappedCoffeeShop(coffeeShop);
        presenter.showDetailView();
    }
    //endregion

    //region ListFragment.Callback
    @Override
    public void onItemTapped(CoffeeShop coffeeShop) {
        presenter.setLastTappedCoffeeShop(coffeeShop);
        presenter.showDetailView();

        MapsFragment mapsFragment = getMapFragment();
        if (mapsFragment != null) {
            mapsFragment.setMarkerActive(coffeeShop);
        }
    }
    //endregion

    @Nullable
    private MapsFragment getMapFragment() {
        Fragment mapFragment = getSupportFragmentManager().findFragmentById(R.id.main_container);
        if (mapFragment instanceof MapsFragment) {
            return (MapsFragment) mapFragment;
        } else {
            return null;
        }
    }
}

File: app/src/main/java/tw/com/louis383/coffeefinder/mainpage/MainActivity.kt
package tw.com.louis383.coffeefinder.mainpage

import android.Manifest
import android.animation.ObjectAnimator
import android.app.Activity
import android.content.Intent
import android.content.pm.PackageManager
import android.content.res.ColorStateList
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.provider.Settings
import android.view.View
import android.view.WindowInsetsController.APPEARANCE_LIGHT_STATUS_BARS
import android.view.animation.AnticipateInterpolator
import android.widget.FrameLayout
import android.widget.ImageButton
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AppCompatActivity
import androidx.coordinatorlayout.widget.CoordinatorLayout
import androidx.core.animation.doOnEnd
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.core.splashscreen.SplashScreen
import androidx.core.splashscreen.SplashScreen.Companion.installSplashScreen
import androidx.core.splashscreen.SplashScreenViewProvider
import androidx.lifecycle.LifecycleOwner
import androidx.viewpager.widget.ViewPager
import com.google.android.gms.common.api.ResolvableApiException
import com.google.android.gms.maps.model.LatLng
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.google.android.material.snackbar.Snackbar
import com.trafi.anchorbottomsheetbehavior.AnchorBottomSheetBehavior
import dagger.hilt.android.AndroidEntryPoint
import tw.com.louis383.coffeefinder.R
import tw.com.louis383.coffeefinder.adapter.ViewPagerAdapter
import tw.com.louis383.coffeefinder.core.domain.model.CoffeeShop
import tw.com.louis383.coffeefinder.details.DetailsFragment
import tw.com.louis383.coffeefinder.details.DetailsItemClickListener
import tw.com.louis383.coffeefinder.list.ListFragment
import tw.com.louis383.coffeefinder.maps.MapsClickHandler
import tw.com.louis383.coffeefinder.maps.MapsFragment
import tw.com.louis383.coffeefinder.uimodel.getUiModel
import tw.com.louis383.coffeefinder.utils.QuickCheckUtils
import tw.com.louis383.coffeefinder.utils.bindView
import javax.inject.Inject

/**
 * Created by louis383 on 2017/2/17.
 */

@AndroidEntryPoint
class MainActivity : AppCompatActivity(), MainView, MapsClickHandler, ListFragment.Callback,
    DetailsItemClickListener, View.OnClickListener {
    private val locationSettingResolution = 1001

    @Inject
    lateinit var presenter: MainPresenter

    private var snackbar: Snackbar? = null

    // Main Content
    private val rootView: CoordinatorLayout by bindView(R.id.main_rootview)
    private val mainContainer: FrameLayout by bindView(R.id.main_container)
    private val myLocationButton: ImageButton by bindView(R.id.main_my_location_button)

    // Bottom Sheet
    private lateinit var bottomSheetViewPager: ViewPager
    private lateinit var bottomSheetBehavior: AnchorBottomSheetBehavior<ViewPager>
    private val viewPagerAdapter = ViewPagerAdapter(supportFragmentManager)

    // View States
    sealed class ViewState {
        data class EnterDetailInfoFromMap(val coffeeshop: CoffeeShop): ViewState()
        data class EnterDetailInfoFromList(val coffeeshop: CoffeeShop): ViewState()
        object Browsing: ViewState()
    }
    private var detailViewState: ViewState = ViewState.Browsing

    private val mapFragment: MapsFragment?
        get() {
            val mapFragment = supportFragmentManager.findFragmentById(R.id.main_container)
            return mapFragment as? MapsFragment
        }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val splashScreen = installSplashScreen()
        handleSplashScreen(splashScreen)
        setContentView(R.layout.activity_main)

        initMapFragment()
        bottomSheetViewPager = findViewById(R.id.main_bottom_sheet)
        bottomSheetViewPager.adapter = viewPagerAdapter
        bottomSheetBehavior = getViewPagerBottomSheetBehavior()
        bottomSheetBehavior.state = AnchorBottomSheetBehavior.STATE_COLLAPSED

        if (QuickCheckUtils.canApplyDynamicColor()) {
            myLocationButton.imageTintList = ColorStateList.valueOf(ContextCompat.getColor(this, R.color.dynamic_light_primary))
        }

        presenter.attachView(this)
        myLocationButton.setOnClickListener(this)
    }

    private fun handleSplashScreen(splashScreen: SplashScreen) {
        splashScreen.setOnExitAnimationListener(object : SplashScreen.OnExitAnimationListener {
            override fun onSplashScreenExit(splashScreenViewProvider: SplashScreenViewProvider) {
                val slideUp = ObjectAnimator.ofFloat(
                    splashScreenViewProvider.view,
                    View.TRANSLATION_Y,
                    0f,
                    splashScreenViewProvider.view.height.toFloat() * -1f
                )
                slideUp.interpolator = AnticipateInterpolator()
                slideUp.duration = 200L

                slideUp.doOnEnd {
                    splashScreenViewProvider.remove()
                }
                slideUp.start()
            }
        })
    }

    private fun initMapFragment() {
        val mapsFragment = MapsFragment.newInstance()
        with(mapsFragment) {
            retainInstance = true
            setMapClickHandler(this@MainActivity)
        }

        supportFragmentManager.beginTransaction()
            .replace(R.id.main_container, mapsFragment, MapsFragment.TAG)
            .commit()
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        when (requestCode) {
            locationSettingResolution -> {
                if (resultCode == Activity.RESULT_OK) {
                    presenter.requestUserLocation()
                } else {
                    snackbar = Snackbar.make(rootView, R.string.high_accuracy_recommand, Snackbar.LENGTH_LONG)
                    snackbar?.show()
                }
            }
            else -> super.onActivityResult(requestCode, resultCode, data)
        }
    }

    override fun provideLifecycleOwner(): LifecycleOwner {
        return this
    }

    private fun forceRequestCoffeeShop() {
        presenter.requestUserLocation()
    }

    override fun checkLocationPermission(): Boolean {
        return ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED ||
                hasApproximateLocationPermission()
    }

    override fun hasApproximateLocationPermission(): Boolean {
        return ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_COARSE_LOCATION) == PackageManager.PERMISSION_GRANTED
    }

    private val locationPermissionCallback = registerForActivityResult(
        ActivityResultContracts.RequestMultiplePermissions()
    ) { permissions ->
        when {
            permissions.getOrDefault(Manifest.permission.ACCESS_FINE_LOCATION, false) -> {
                // Precise location access granted.
                presenter.requestUserLocation()
                hideSnackBar()
            }
            permissions.getOrDefault(Manifest.permission.ACCESS_COARSE_LOCATION, false) -> {
                // Only approximate location access granted.
                presenter.requestUserLocation()
                hideSnackBar()
            }
            else -> {
                // No location access granted.
                if (ActivityCompat.shouldShowRequestPermissionRationale(this, Manifest.permission.ACCESS_COARSE_LOCATION) &&
                    ActivityCompat.shouldShowRequestPermissionRationale(this, Manifest.permission.ACCESS_FINE_LOCATION)) {
                    MaterialAlertDialogBuilder(this)
                        .setMessage(getResourceString(R.string.request_location))
                        .setPositiveButton(getResourceString(R.string.dialog_auth))
                        { _, _ -> presenter.requestLocationPermission() }
                        .setNegativeButton(getResourceString(R.string.dialog_cancel))
                        { _, _ -> showPermissionNeedSnackBar() }
                        .create()
                        .show()
                } else {
                    showPermissionNeedSnackBar()
                    val appName = getResourceString(R.string.app_name)
                    val permissionName = getResourceString(R.string.auth_location)

                    MaterialAlertDialogBuilder(this)
                        .setTitle(getResourceString(R.string.dialog_auth))
                        .setMessage(resources.getString(R.string.auth_yourself, appName, permissionName))
                        .setPositiveButton(getResourceString(R.string.auto_go))
                        { _, _ -> openApplicationSetting() }
                        .setNegativeButton(getResourceString(R.string.dialog_cancel))
                        { _, _ -> }
                        .create()
                        .show()
                }
            }
        }
    }

    override fun requestLocationPermission() {
        locationPermissionCallback.launch(
            arrayOf(
                Manifest.permission.ACCESS_FINE_LOCATION,
                Manifest.permission.ACCESS_COARSE_LOCATION
            )
        )
    }

    override fun locationSettingNeedsResolution(resolvable: ResolvableApiException) {
        resolvable.startResolutionForResult(this, locationSettingResolution)
    }

    override fun showServiceUnavailableMessage() {
        val message = resources.getString(R.string.service_unavailable)
        makeSnackBar(message, true)
    }

    override fun makeSnackBar(message: String, infinity: Boolean) {
        val duration = if (infinity) Snackbar.LENGTH_INDEFINITE else Snackbar.LENGTH_LONG
        val snackbar = Snackbar.make(rootView, message, duration)
        snackbar.show()
        this.snackbar = snackbar
    }

    override fun makeSnackBar(stringRes: Int) {
        if (stringRes != -1) {
            val message = getString(stringRes)
            makeSnackBar(message, false)
        }
    }

    override fun setStatusBarDarkIndicator() {
        translucentStatusBar()
    }

    private fun translucentStatusBar() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            // replace View.SYSTEM_UI_FLAG_LAYOUT_STABLE or View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
            window.setDecorFitsSystemWindows(false)
            // replace View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR
            window.insetsController?.setSystemBarsAppearance(APPEARANCE_LIGHT_STATUS_BARS, APPEARANCE_LIGHT_STATUS_BARS)
        } else {
            window.decorView.systemUiVisibility = View.SYSTEM_UI_FLAG_LAYOUT_STABLE or
                    View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN or
                    View.SYSTEM_UI_FLAG_LIGHT_STATUS_BAR
        }
    }

    private val internetRequestCallback = registerForActivityResult(
        ActivityResultContracts.StartActivityForResult()
    ) {
        val isNetworkAvailable = presenter.isNetworkAvailable()
        if (isNetworkAvailable) {
            if (checkLocationPermission()) {
                forceRequestCoffeeShop()
            } else {
                showPermissionNeedSnackBar()
            }
        } else {
            requestInternetConnection()
        }
    }

    override fun requestInternetConnection() {
        MaterialAlertDialogBuilder(this)
            .setTitle(getResourceString(R.string.internet_request_title))
            .setMessage(getResourceString(R.string.internet_request_message))
            .setPositiveButton(getResourceString(R.string.auto_go)) { _, _ ->
                internetRequestCallback.launch(Intent(Settings.ACTION_WIFI_SETTINGS))
            }.create().show()
    }

    override fun moveCameraToCurrentPosition(latLng: LatLng) {
        mapFragment?.moveCamera(latLng, MapsFragment.ZOOM_RATE)
    }

    override fun onCoffeeShopFetched(coffeeShops: List<CoffeeShop>) {
        mapFragment?.prepareCoffeeShops(coffeeShops)
    }

    override fun updateListPage(coffeeShops: List<CoffeeShop>) {
        if (!viewPagerAdapter.isListPageInitiated) {
            val listFragment = ListFragment.newInstance(coffeeShops)
            listFragment.setCallback(this)
            // FIXME:: Below line cause crash
            viewPagerAdapter.setListFragment(listFragment)
        }
    }

    override fun isApplicationInstalled(packageName: String): Boolean {
        val packageManager = packageManager
        return try {
            packageManager.getPackageInfo(packageName, PackageManager.GET_ACTIVITIES)
            true
        } catch (e: PackageManager.NameNotFoundException) {
            false
        }
    }

    override fun navigateToLocation(intent: Intent) {
        startActivity(intent)
    }

    override fun showNeedsGoogleMapMessage() {
        val message = resources.getString(R.string.googlemap_not_install)
        makeSnackBar(message, false)
    }

    override fun shareCoffeeShop(shareIntent: Intent) {
        val title = getResourceString(R.string.share_title)
        startActivity(Intent.createChooser(shareIntent, title))
    }

    private fun showPermissionNeedSnackBar() {
        val snackbar = Snackbar.make(rootView, R.string.permission_needed, Snackbar.LENGTH_INDEFINITE)
        snackbar.setAction(R.string.dialog_auth) { openApplicationSetting() }
        snackbar.show()
        this.snackbar = snackbar
    }

    private val enableLocationManuallyCallback = registerForActivityResult(
        ActivityResultContracts.StartActivityForResult()
    ) {
        if (checkLocationPermission()) {
            forceRequestCoffeeShop()
            hideSnackBar()
        } else {
            showPermissionNeedSnackBar()
        }
    }

    private fun hideSnackBar() {
        snackbar?.run {
            if (isShown) {
                dismiss()
            }
        }
    }

    private fun openApplicationSetting() {
        val intent = Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS).apply {
            data = Uri.fromParts("package", packageName, null)
        }
        enableLocationManuallyCallback.launch(intent)
    }

    private fun getResourceString(stringId: Int): String {
        return resources.getString(stringId)
    }

    override fun showBottomSheetDetailView(coffeeShop: CoffeeShop) {
        if (this::bottomSheetViewPager.isInitialized) {
            if (viewPagerAdapter.isDetailPageInitiated) {
                val detailsFragment = viewPagerAdapter.getItem(ViewPagerAdapter.DETAIL_FRAGMENT)
                (detailsFragment as DetailsFragment).setDetailInfo(coffeeShop.getUiModel())
            } else {
                val detailsFragment = DetailsFragment.newInstance(coffeeShop)
                detailsFragment.detailsItemClickListener = this
                viewPagerAdapter.setDetailFragment(detailsFragment)
            }

            if (bottomSheetBehavior.state != AnchorBottomSheetBehavior.STATE_ANCHORED) {
                bottomSheetBehavior.state = AnchorBottomSheetBehavior.STATE_ANCHORED
            }

            listScrollToItemPosition(coffeeShop)
            bottomSheetViewPager.setCurrentItem(ViewPagerAdapter.DETAIL_FRAGMENT, true)
        }
    }

    private fun listScrollToItemPosition(coffeeShop: CoffeeShop) {
        if (viewPagerAdapter.isListPageInitiated) {
            val listFragment = viewPagerAdapter.getItem(ViewPagerAdapter.LIST_FRAGMENT)
            (listFragment as ListFragment).scrollToItemPosition(coffeeShop)
        }
    }

    override fun moveMapView(offset: Float) {
        mainContainer.translationY = offset
    }

    override fun onBackPressed() {
        if (this::bottomSheetBehavior.isInitialized) {
            if (bottomSheetViewPager.currentItem != 0) {
                bottomSheetViewPager.currentItem = 0

                when (detailViewState) {
                    is ViewState.EnterDetailInfoFromList -> {
                        if (bottomSheetBehavior.state != AnchorBottomSheetBehavior.STATE_ANCHORED) {
                            bottomSheetBehavior.state = AnchorBottomSheetBehavior.STATE_ANCHORED
                        }
                    }
                    is ViewState.EnterDetailInfoFromMap -> {
                        if (bottomSheetBehavior.state != AnchorBottomSheetBehavior.STATE_COLLAPSED) {
                            bottomSheetBehavior.state = AnchorBottomSheetBehavior.STATE_COLLAPSED
                        }
                    }
                    else -> {}
                }
                detailViewState = ViewState.Browsing
                return
            }
        }
        super.onBackPressed()
    }

    override fun getViewPagerBottomSheetBehavior(): AnchorBottomSheetBehavior<ViewPager> = AnchorBottomSheetBehavior.from(bottomSheetViewPager)

    //region MapsClickHandler
    override fun onMapClicked() {
        if (this::bottomSheetBehavior.isInitialized) {
            if (bottomSheetBehavior.state != AnchorBottomSheetBehavior.STATE_COLLAPSED) {
                bottomSheetBehavior.state = AnchorBottomSheetBehavior.STATE_COLLAPSED
            }
        }
    }

    override fun onMarkerClicked(coffeeShop: CoffeeShop) {
        presenter.setLastTappedCoffeeShop(coffeeShop)
        presenter.showDetailView()
        detailViewState = ViewState.EnterDetailInfoFromMap(coffeeShop)
    }
    //endregion

    //region ListFragment.Callback
    override fun onItemTapped(coffeeShop: CoffeeShop) {
        presenter.setLastTappedCoffeeShop(coffeeShop)
        presenter.showDetailView()
        detailViewState = ViewState.EnterDetailInfoFromList(coffeeShop)
        mapFragment?.setMarkerActive(coffeeShop)
    }
    //endregion

    //region DetailsItemClickListener
    override fun onNavigationButtonClicked() {
        presenter.prepareNavigation()
    }

    override fun onShareButtonClicked() {
        presenter.share(this)
    }

    override fun onBackButtonClicked() {
        onBackPressed()
    }
    //endregion

    //region View.OnClickListener
    override fun onClick(view: View) {
        when (view.id) {
            R.id.main_my_location_button -> {
                presenter.let {
                    val myLocation = it.currentLocation
                    val mapFragment = supportFragmentManager.findFragmentByTag(MapsFragment.TAG)
                            as MapsFragment
                    mapFragment.moveToMyLocation(myLocation)
                }
            }
        }
    }
    //endregion
}
